src/github.com/matrix-org/dendrite/syncapi/storage/syncserver.go

```diff
@@ -323,6 +324,49 @@ func (d *SyncServerDatabase) fetchStateEvents(txn *sql.Tx, roomIDToEventIDSet ma
 	return stateBetween, nil
 }
 
+func (d *SyncServerDatabase) fetchMissingStateEvents(txn *sql.Tx, eventIDs []string) ([]streamEvent, error) {
+	// Fetch from the events table first so we pick up the stream ID for the
+	// event.
+	events, err := d.events.selectEvents(txn, eventIDs)
+	if err != nil {
+		return nil, err
+	}
+
+	have := map[string]bool{}
+	for _, event := range events {
+		have[event.EventID()] = true
+	}
+	var missing []string
+	for _, eventID := range eventIDs {
+		if !have[eventID] {
+			missing = append(missing, eventID)
+		}
+	}
+	if len(missing) == 0 {
+		return events, nil
+	}
+
+	// If they are missing from the events table then they should be state
+	// events that we received from outside the main event stream.
+	// These should be in the room state table.
+	stateEvents, err := d.roomstate.selectEventsWithEventIDs(txn, missing)
+
+	if err != nil {
+		return nil, err
+	}
+	if len(stateEvents) != len(missing) {
+		return nil, fmt.Errorf("failed to map all event IDs to events: (got %d, wanted %d)", len(stateEvents), len(missing))
+	}
+	for _, e := range stateEvents {
+		// Set the stream position to 0 since these events occured outside the
+		// stream so probably happened before it.
+		// TOOD: What happens if we receive a state event from outside the
+		// timeline associated with an event in the middle of the timeline?
```

Comment:
> TOOD

I've mulled over this. This scenario is hit when we get arbitrary state events from federation. It is the room servers job to keep a consistent current room state based on deltas to add/remove state event IDs. The room server has instructed us that this forms part of the current state and therefore should be in the current state table. However, the room server has not told us which stream position the event should be in. 

We're *assuming* they happened outside the stream (backfill from federation or equiv) and hence setting the stream position to 0. This may change when we eventually implement backfill and realise it should be `-435` or whatever. I think this behaviour is probably okay for now, but will need to be revisited when we implement backfill.